File: apps/api/app/schemas/privacy.py
from __future__ import annotations

from datetime import datetime
from pathlib import Path
from typing import Literal

from pydantic import Field, SecretStr

from app.schemas.common import ApiModel


class BackupFileEntry(ApiModel):
    path: str = Field(min_length=1, max_length=1000)
    kind: Literal["attachment", "database", "preferences"]
    size_bytes: int = Field(ge=0)
    sha256: str = Field(pattern=r"^[a-f0-9]{64}$")


class BackupManifest(ApiModel):
    format: Literal["locallife-backup"] = "locallife-backup"
    format_version: Literal[1] = 1
    app_version: str
    created_at: datetime
    schema_revision: str
    encrypted: bool
    database_path: Literal["database/locallife.db"] = "database/locallife.db"
    preferences_path: Literal["preferences.json"] = "preferences.json"
    workspace_metadata: dict[str, str]
    preference_metadata: dict[str, str | int]
    files: list[BackupFileEntry]


class BackupSummary(ApiModel):
    filename: str
    path: Path
    created_at: datetime
    schema_revision: str
    encrypted: bool
    size_bytes: int = Field(ge=0)
    integrity_verified: bool


class BackupCreateRequest(ApiModel):
    password: SecretStr | None = None
    label: str | None = Field(default=None, min_length=1, max_length=40)


class BackupCreateResponse(ApiModel):
    backup: BackupSummary
    manifest: BackupManifest


class PrivacyStatusResponse(ApiModel):
    data_directory: Path
    database_path: Path
    attachments_directory: Path
    backups_directory: Path
    imports_directory: Path
    network_mode: Literal["loopback-only"]
    telemetry_enabled: Literal[False]
    external_requests_enabled: bool
    outbound_guard_active: bool
    max_attachment_bytes: int
    max_import_bytes: int
    max_backup_bytes: int
    session_timeout_minutes: int
    privacy_lock_scope: Literal["casual-screen-privacy"]
    last_backup: BackupSummary | None


class DeleteAllLocalDataRequest(ApiModel):
    confirmation: Literal["DELETE ALL LOCAL DATA"]
    include_backups: bool = False


class DeleteAllLocalDataResponse(ApiModel):
    deleted_database_records: int
    deleted_attachment_files: int
    deleted_import_files: int
    deleted_backup_files: int
    preserved_backups: bool

File: apps/web/lib/api/privacy.ts
import { apiRequest, jsonBody } from "./client";
import type { DataEnvelope } from "./types";

export interface BackupSummary {
  filename: string;
  path: string;
  created_at: string;
  schema_revision: string;
  encrypted: boolean;
  size_bytes: number;
  integrity_verified: boolean;
}

export interface PrivacyStatus {
  data_directory: string;
  database_path: string;
  attachments_directory: string;
  backups_directory: string;
  imports_directory: string;
  network_mode: "loopback-only";
  telemetry_enabled: false;
  external_requests_enabled: boolean;
  outbound_guard_active: boolean;
  max_attachment_bytes: number;
  max_import_bytes: number;
  max_backup_bytes: number;
  session_timeout_minutes: number;
  privacy_lock_scope: "casual-screen-privacy";
  last_backup: BackupSummary | null;
}

export interface BackupCreateResult {
  backup: BackupSummary;
  manifest: {
    format: "locallife-backup";
    format_version: 1;
    encrypted: boolean;
    files: Array<{ path: string; size_bytes: number; sha256: string }>;
  };
}

export interface DeleteAllResult {
  deleted_database_records: number;
  deleted_attachment_files: number;
  deleted_import_files: number;
  deleted_backup_files: number;
  preserved_backups: boolean;
}

export async function getPrivacyStatus(): Promise<PrivacyStatus> {
  return (await apiRequest<DataEnvelope<PrivacyStatus>>("/privacy/status")).data;
}

export async function createLocalBackup(payload: {
  label?: string;
  password?: string;
}): Promise<BackupCreateResult> {
  return (
    await apiRequest<DataEnvelope<BackupCreateResult>>("/privacy/backups", {
      method: "POST",
      ...jsonBody(payload),
    })
  ).data;
}

export async function deleteAllLocalData(payload: {
  confirmation: "DELETE ALL LOCAL DATA";
  include_backups: boolean;
}): Promise<DeleteAllResult> {
  return (
    await apiRequest<DataEnvelope<DeleteAllResult>>("/privacy/delete-all", {
      method: "POST",
      ...jsonBody(payload),
    })
  ).data;
}
